src/Database/LSMTree/Internal/TableHandle.hs

```diff
@@ -0,0 +1,183 @@
+{-# OPTIONS_GHC -Wno-unused-top-binds #-}
+
+module Database.LSMTree.Internal.TableHandle (
+    -- * Exceptions
+    LSMTreeError (..)
+    -- * Session
+  , Session (..)
+    -- * Table handle
+  , TableHandle
+    -- * configuration
+  , TableConfig (..)
+  , ResolveMupsert (..)
+  , SizeRatio (..)
+  , MergePolicy (..)
+  , BloomFilterAlloc (..)
+  ) where
+
+import           Control.Concurrent.Class.MonadMVar.Strict (StrictMVar)
+import           Control.Exception (Exception)
+import           Data.BloomFilter (Bloom)
+import qualified Data.Vector as V
+import           Data.Word (Word32)
+import           Database.LSMTree.Internal.IndexCompact (IndexCompact)
+import           Database.LSMTree.Internal.Run (Run)
+import qualified Database.LSMTree.Internal.Run as Run
+import           Database.LSMTree.Internal.Serialise (SerialisedKey,
+                     SerialisedValue)
+import           Database.LSMTree.Internal.WriteBuffer (WriteBuffer)
+import           System.FS.API (FsErrorPath, FsPath, HasFS)
+import           System.FS.BlockIO.API (HasBlockIO)
+
+{-------------------------------------------------------------------------------
+  Exceptions
+-------------------------------------------------------------------------------}
+
+-- TODO: give this a nicer Show instance.
+data LSMTreeError =
+    SessionDirDoesNotExist FsErrorPath
+  | SessionDirLocked FsErrorPath
+  deriving (Show, Exception)
+
+{-------------------------------------------------------------------------------
+  Session
+-------------------------------------------------------------------------------}
+
+-- | A session provides context that is shared across multiple table handles.
+--
+-- For more information, see 'Database.LSMTree.Common.Session'.
+data Session m h = Session {
+    -- | The path to the directory in which this sesion is live. This is a path
+    -- relative to root of the 'HasFS' instance.
+    sessionRoot        :: !FsPath
+  , sessionHasFS       :: !(HasFS m h)
+  , sessionHasBlockIO  :: !(HasBlockIO m h)
+    -- | A session-wide shared, atomic counter that is used to produce unique
+    -- names for run files.
+  , sessionUniqCounter :: !(StrictMVar m Integer)
+  }
+
+{-------------------------------------------------------------------------------
+  Table handle
+-------------------------------------------------------------------------------}
+
+-- | A handle to an on-disk key\/value table.
+--
+-- For more information, see 'Database.LSMTree.Normal.TableHandle'.
+data TableHandle m h = TableHandle {
+    tableSession     :: !(Session m h)
+  , tableConfig      :: !TableConfig
+  , tableWriteBuffer :: !(WriteBuffer)
+    -- | A hierarchy of levels. The vector indexes double as level numbers.
+  , tableLevels      :: !(V.Vector (Level h))
+    -- | Cache of flattened 'levels'.
+    --
+    -- INVARIANT: when 'level's is modified, this cache should be updated as
+    -- well, for example using 'mkLevelsCache'.
+  , tableLevelsCache :: !(LevelsCache h)
```

Comment:
Once we have updates, these fields will need to allow for mutation. Also, we will need a synchronisation mechanism. We can leave that for later, but at least a comment might be a good idea.

Reply:
I've added a TODO about a synchronisation mechanism

I think that for these fields, mutation isn't worth it. The levels/cache will rarely change (only when merges are finished) and at that point it's relatively cheap to allocate a new vector where necessary. And then we get sharing for free

Reply:
Or maybe I'm forgetting about a case where we would need these vectors to be mutable?

Reply:
IIRC we decided that table handles would not have a locking mechanism to make using them thread-safe, so I've made the mutable fields into MVars, and added a note about thread safety